q2_annotate/kraken2/merge.py

```diff
@@ -0,0 +1,420 @@
+        The merged kraken2 report format.
+    '''
+    if len(reports) == 1:
+        return reports[0]
+
+    merged: tuple[TreeNode | None, TreeNode | None] | None = None
+    while reports:
+        first = _report_df_to_tree(reports.pop().view(pd.DataFrame))
+        if merged is None:
+            second = _report_df_to_tree(reports.pop().view(pd.DataFrame))
+        else:
+            second = merged
+
+        merged = _merge_trees(first, second)
+
+    merged_report_df = _dump_tree_to_report(*merged)
+    merged_report = Kraken2ReportFormat()
+    merged_report_df.to_csv(
+        str(merged_report), sep='\t', header=False, index=False
+    )
+
+    return merged_report
+
+
+def _merge_outputs(
+    outputs: list[Kraken2OutputFormat]
+) -> Kraken2OutputFormat:
+    '''
+    Merges two or more kraken2 outputs into a single output. Output files
+    are merged by concatenation.
+
+    Parameters
+    ----------
+    outputs : list[Kraken2OutputFormat]
+        Kraken2 outputs belonging to the same sample ID.
+
+    Returns
+    -------
+        The merged kraken2 output format.
+    '''
+    if len(outputs) == 1:
+        return outputs[0]
+
+    merged_output = Kraken2OutputFormat()
+    while outputs:
+        with (
+            open(str(merged_output), 'a') as merged_fh,
+            open(str(outputs.pop()), 'r') as fh
+        ):
+            while buffer := fh.read(4096):
+                merged_fh.write(buffer)
+
+    return merged_output
+
+
+def _merge_trees(
+    first: tuple[TreeNode | None, TreeNode | None],
+    second: tuple[TreeNode | None, TreeNode | None]
+) -> tuple[TreeNode | None, TreeNode | None]:
+    '''
+    Merges two trees each representing a kraken2 report into a single tree.
+    The number of reads assigned to each node are summed where nodes overlap,
+    and new nodes are inserted into the tree where they don't. The proportions
+    of assigned and covered reads are then updated in a final passover.
+
+    Parameters
+    ----------
+    first : tuple[TreeNode | None, TreeNode | None]
+        The first report tree, where the first node in the tuple represents the
+        tree and the second node represents an optional unclassified node.
+    second : tuple[TreeNode | None, TreeNode | None]
+        The second report tree, where the first node in the tuple represents the
+        tree and the second node represents an optional unclassified node.
+
+    Returns
+    -------
+    tuple[TreeNode | None, TreeNode | None]
+        The merged tree.
+    '''
+    first_tree, first_unclassified_node = first
+    second_tree, second_unclassified_node = second
+
+    # merge trees (with respect to `n_frags_assigned`, `n_frags_covered`)
+    if first_tree is None and second_tree is None:
+        merged_tree = None
+    elif first_tree is None:
+        merged_tree = second_tree
+    elif second_tree is None:
+        merged_tree = first_tree
+    else:
+        for node in first_tree.levelorder():
```

Comment:
Adding comments to the individual steps your taking in here (i.e., why are you adding the specific values that you're adding; why is it same to assume a node's parent from `first` is in `second`, ...) would be helpful for interpreting this section.